### PR TITLE
add Debian packaging for python-twodict

### PR DESCRIPTION
**Pull request by nodiscc**
Related to https://github.com/MrS0m30n3/youtube-dl-gui/pull/189

The best and recommended practice is to only push the `debian/` directory to a separate branch (eg. `debian-packaging`), so that it does not interfere with your master branch. When you want to release a new package, just `git checkout debian-packaging; git merge master`, bump the package version/changelog/etc and rebuild.

But Github doesn't let me create a Pull Request for a _new branch_, so you'd have to merge this manually:

```
git remote add nodiscc https://github.com/nodiscc/twodict
git remote update
git checkout debian-packaging
git push -u origin debian-packaging
```

If you want to maintain the .deb packages in the future (I don't know if/how much you use Debian) You may also want to edit the `Maintainer:` field in debian/control. You should also create a GnuPG key for package signing. Otherwise let me know and I'll try to keep them updated.

### CODE REVIEW

**Comment by nodiscc** [comment]
@MrS0m30n3 ok to merge?

**Comment by nodiscc** [comment]
@MrS0m30n3 I can also move _everything_ (including twodict packaging) to a `packaging/debian/` directory in https://github.com/MrS0m30n3/youtube-dl-gui/ if you don't want debian directories in all your projects. Just tell me what you prefer and I'll adapt the patch.

File: debian/control [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Source: python-twodict
+Section: python
+Priority: optional
+Maintainer: nodiscc <nodiscc@gmail.com>
+Build-Depends: debhelper (>= 9), dh-python, python-all
```
Comment (MrS0m30n3) [inline debian/control]: You should also add `python3-all` since twodict supports Python 3.*

File: debian/control [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+X-Python-Version: >= 2.7
```
Comment (MrS0m30n3) [inline debian/control]: You should also add `X-Python3-Version: >= 3.3`

File: debian/control [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+Standards-Version: 3.9.8
+Homepage: https://github.com/MrS0m30n3/twodict
+
+Package: python-twodict
+Architecture: all
+Depends: ${shlibs:Depends}, ${misc:Depends}, ${python:Depends}
```
Comment (MrS0m30n3) [inline debian/control]: \+ `${python3:Depends}`

File: debian/rules [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/make -f
+
+%:
+	dh $@ --with python2
```
Comment (MrS0m30n3) [inline debian/rules]: Use **pybuild**, check [here](https://wiki.debian.org/Python/LibraryStyleGuide?action=show&redirect=Python#debian.2Frules)

Comment (MrS0m30n3) [reply debian/rules]: Also, make sure the tests pass both in Python2 & 3

**Review by MrS0m30n3 — COMMENTED** [review]
Excellent job @nodiscc, you should also add support for Python3 since twodict supports both Python2 & 3.

**Comment by MrS0m30n3** [comment]
Regarding the **debian** directory i have no problem to keep it the main directory as long as we use a different branch.

**Comment by MrS0m30n3** [comment]
I also created a new branch `debian-packaging`. Now you should be able to push your changes there.

**Comment by nodiscc** [comment]
Edit: fixed

~~Thanks, I made the required changes in https://github.com/MrS0m30n3/twodict/pull/1/commits/00fee07219970097a760d51881a5a5082433bc17 , but I am getting warnings `unknown substitution variable ${...:Depends}`, there is no `python3-twodict` package produced, and lintian complains about the python2 package. I'm attaching the log, at this point I have no clue but will investigate, please let me know if you have more info on how to fix this~~

**Comment by nodiscc** [comment]
My bad, I actually had edited the wrong target in `debian/rules`. Building the package seems to work, no `python3-twodict` package is produced, I will check if this is expected. The following lintian warnings have appeared, I will try to deal with them ASAP

```
E: python-twodict: package-installs-python-pycache-dir usr/lib/python3.5/dist-packages/__pycache__/
W: python-twodict: python-module-in-wrong-location usr/lib/python3.5/dist-packages/__pycache__/ usr/lib/python3/dist-packages/__pycache__/
W: python-twodict: python-module-in-wrong-location usr/lib/python3.5/dist-packages/__pycache__/twodict.cpython-35.pyc usr/lib/python3/dist-packages/__pycache__/twodict.cpython-35.pyc
W: python-twodict: python-module-in-wrong-location usr/lib/python3.5/dist-packages/twodict-1.2.egg-info usr/lib/python3/dist-packages/twodict-1.2.egg-info
W: python-twodict: python-module-in-wrong-location ... use --no-tag-display-limit to see all (or pipe to a file/program)
```

**Comment by MrS0m30n3** [comment]
`package-installs-python-pycache-dir` the main reason for this (probably), is that twodict uses the `distutils` module instead of the newer `setuptools` in the **setup.py** file.

I guess we could add a rule in the `debian/rules` file to remove the **\_\_pycache\_\_** dir before packaging.

**Comment by nodiscc** [comment]
https://lintian.debian.org/tags/package-installs-python-pycache-dir.html:
> The package installs a `__pycache__` directory, which is normally only used to store compiled Python source files. Compiled Python source files must not be included in the package, instead they should be generated at installation time in the postinst.

I'll try to find a similar debian package to see how this is done.

**Comment by MrS0m30n3** [comment]
Adding the rule below in **debian/rules** seems to work. Not sure if this is the proper way to do it though.

```make
override_dh_lintian:
    rm -rfv debian/python-twodict/usr/lib/python3.5/dist-packages/__pycache__/
    dh_lintian
```

```
$ lintian python-twodict_1.2_all.deb 
W: python-twodict: python-module-in-wrong-location usr/lib/python3.5/dist-packages/twodict-1.2.egg-info usr/lib/python3/dist-packages/twodict-1.2.egg-info
W: python-twodict: python-module-in-wrong-location usr/lib/python3.5/dist-packages/twodict.py usr/lib/python3/dist-packages/twodict.py
```

**Comment by legrostdg** [comment]
Damn... https://salsa.debian.org/felix-guest/twodict
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=886730

**Comment by MrS0m30n3** [comment]
Do the tests actually pass?? When i run `dpkg-buildpackage -rfakeroot` i get:
```
...
Ran 0 test in 0.000s
...
```

**Comment by legrostdg** [comment]
They do, in my packaging. The package is in the NEW queue, so it should be available relatively soon in Debian unstable.

**Comment by MrS0m30n3** [comment]
What commands you use to build the package?

**Comment by legrostdg** [comment]
gbp buildpackage

**Comment by MrS0m30n3** [comment]
Again, the tests won't run for me `Ran 0 test...`.
Are you sure **all** tests run?

**Comment by legrostdg** [comment]
You're sure you used  https://salsa.debian.org/felix-guest/twodict.git ?

**Comment by legrostdg** [comment]
Woops, I meant https://anonscm.debian.org/git/python-modules/packages/twodict.git

**Comment by legrostdg** [comment]
(I had to move the repo there to be hosted by the Debian Python Packaging Team, but the first repo should be uptodate, too)

**Comment by legrostdg** [comment]
I'll just remove the repo on salsa to avoid confusion.

**Comment by MrS0m30n3** [comment]
I was using the repo on *salsa*, that was the problem. On the *anonscm* repo everything runs fine.

**Comment by MrS0m30n3** [comment]
Anyway, closing. Great job to both of you!

**Pull request closed without merging**